cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+1. If the TLS handshake succeeds, then the connection is authenticated and the traffic protected by policy can proceed. The TLS connection between the agents can then be torn down and thrown away (since it’s not being used to tunnel the encrypted traffic through).
+
+
+### Control Plane
+
+#### Structures
+
+Auth table - a BPF table needs to be added that records, at least:
+
+* Source Identity
+* Destination Identity
+* Destination Node
+* Authentication expiry time
+
+Further details of this table will be worked out as part of the implementation.
+
+#### Connections
+
+TODO: This section needs updating
+
+##### Between Cilium Operator instances and the SPIFFE server
+
+SPIFFE identity registration must occur out-of-band from when the actual SPIFFE identity usage is made via the workload API. In order to register the SPIFFE identities, the Cilium Operator must share access to an admin domain socket with the SPIRE Server, and use that domain socket for creation of new SPIFFE identities in the SPIRE server.
+
+Notably, this means that _one_ Pod out of the Operator deployment must be running on the _same node_ as the SPIRE Server's StatefulSet Pod. This will need Pod Affinity, and some way for the Operator to check if the domain socket is available. This process will also need to run independently of the usual Kubernetes Leadership, unfortunately.
+
+##### Between Cilium Agent instance peers
+Mutual authentication connections are established between the node IPs themselves. However, the peers that require authentication will not be these nodes.
+
+The source agent determines the destination agent using the destination endpoint IP. It then reaches out and connects to the listening port on the destination agent, and completes a TLS handshake. After the TLS handshake is complete, the session is torn down, as all the relevant information has been required. (Namely, did the authentication work, and when do the client and server expire?)
+
+This design has a few wins here:
+- The Authentication only requires that the TLS Handshake succeed, it does not require that we keep the connection open, so we will be making short lived connections for each authentication check, and this will most likely need to be to a new, dedicated port.
+- Because we will be controlling both ends of the TLS handshake, we can use any field we like in the certificates for authentication - SPIFFE certificates issued by SPIRE have the full SPIFFE ID URI as a SAN field, so we can use that for choosing both client and server certificates - the TLS request from the client will be requesting the full SPIFFE ID as its URI.
```

Comment:
https://datatracker.ietf.org/doc/html/rfc6066#section-3 the SNI can only be a DNS name and not a full URL. However we can send the Cilium ID as it's hostname and control it that way

Reply:
hmm, right. Maybe we may need to settle on hostname standard we can use for these requests, something like `<sourceidnumber><destinationidnumber>.<trustdomain>` or something.

Reply:
I've updated, thanks!